Doc/1114/explain prefabs (#1340)
amethyst/amethyst: book/src/prefabs/prefabs_technical_explanation.md

# Prefabs Technical Explanation

A `Prefab` in Amethyst is at the core a simple list of future entities, where each entry in
the list consists of two pieces of optional data:

* a parent index that refers to a different entry in the list
* a data collection implementing the trait `PrefabData`

> To instantiate a `Prefab`, we put a `Handle<Prefab<T>>` on an `Entity`. The `Entity` we put

Reviewer: This handle? https://www.amethyst.rs/doc/master/doc/amethyst_assets/struct.Handle.html
Maybe specify which handle is the right? (searching "Handle" in the doc gives 179 results)

the `Handle` on is referred to as the main `Entity`, and the first entry in the list inside a

> `Prefab` refers to this `Entity`. All other entries in the list will spawn a new `Entity` on

Reviewer: ```suggestion
`Prefab` will be applied to this `Entity`. All other entries in the list will spawn a new `Entity` on
```
I didn't understand on first read and had to check the documentation. What about this?

Author: Oh, this was a pre existing file that I renamed (no content was changed), I'm planning to break it up into separate "how to" guides in the coming weeks

A `PrefabData` implementation could refer to other assets that need to be loaded asynchronously, and
we don't want the user get a `Complete` notification on their `Progress` before everything has been
loaded.

Because of this, once the `Format` have loaded the `Prefab` from the `Source`, and a `PrefabLoaderSystem`
runs `process` on the `AssetStorage`, the system will invoke the `load_sub_assets` function on the
`PrefabData` implementation. If any asset loads are triggered during this, they must adhere to the following
rules:

* the given `ProgressCounter` must be used as a parameter to the load function on `Loader`, so load tracking
works correctly
* the function must return `Ok(true)` (unless an `Error` occurred)

Note that during this phase the `PrefabData` is mutable, which means it can morph inside the `Prefab`. An
example of this is the `AssetPrefab`, which will morph into `AssetPrefab::Handle`.

Once all sub asset loading is finished, which the `PrefabLoaderSystem` will track using the `ProgressCounter`,
a `Complete` signal will be sent upwards.

**Prefab instantiation**

This stage happens after the `Prefab` has been fully loaded and `Complete` has been signaled, and the

> `Handle<Prefab<T>>` is put on an `Entity`. At this point we know that all internal data has been loaded,

Reviewer: Is that the main entity?

Note that for prefabs that reference other prefabs, to make instantiation be performed inside a single frame,
lower level `PrefabLoaderSystem`s need to depend on the higher level ones. To see how this works out check the gltf
example, where we have a scene prefab, and the gltf loader (which use the prefab system internally).

## `PrefabData`

Ok, so what would a simple implementation of `PrefabData` look like?

Let's take a look at the implementation for `Transform`, which is a core concept in Amethyst:

```rust,no_run,noplaypen
# extern crate amethyst;
# use amethyst::assets::PrefabData;
# use amethyst::ecs::{WriteStorage, Entity, Component, NullStorage, error::Error as SpecsError};
#
# // We declare that struct for the sake of automated testing.
# #[derive(Default, Clone)]
# struct Transform;
# impl Component for Transform {
#   type Storage = NullStorage<Transform>;
# }
#
impl<'a> PrefabData<'a> for Transform {
    type SystemData = WriteStorage<'a, Transform>;
    type Result = ();

    fn add_to_entity(
        &self,
        entity: Entity,
        storage: &mut Self::SystemData,
        _: &[Entity],
    ) -> Result<(), SpecsError> {

>         storage.insert(entity, self.clone()).map(|_| ())

Reviewer: ```suggestion
        storage.insert(entity, self.clone());
        Ok(())
```
What do you think of that? This may be easier to understand.

    }
}
```

First, we specify a `SystemData` type, this is the data required from `World` in order to load and
instantiate this `PrefabData`. Here we only need to write to `Transform`.

Second, we specify what result the `add_to_entity` function returns. In our case this is unit `()`, for

> other implementations it could return a `Handle` etc. For an example of this, look at the `TexturePrefab`

Reviewer: The next example is also an example of that.

    ) -> Result<Handle<A>, SpecsError> {
        let handle = match *self {
            AssetPrefab::Handle(ref handle) => handle.clone(),
            AssetPrefab::File(ref name, ref format, ref options) => system_data.0.load(
                name.as_ref(),
                format.clone(),
                options.clone(),
                (),
                &system_data.2,
            ),
        };
        system_data.1.insert(entity, handle.clone()).map(|_| handle)
    }

    fn load_sub_assets(
        &mut self,
        progress: &mut ProgressCounter,
        system_data: &mut Self::SystemData,
    ) -> Result<bool, SpecsError> {
        let handle = match *self {
            AssetPrefab::File(ref name, ref format, ref options) => Some(system_data.0.load(
                name.as_ref(),
                format.clone(),
                options.clone(),
                progress,
                &system_data.2,
            )),
            _ => None,
        };
        if let Some(handle) = handle {
            *self = AssetPrefab::Handle(handle);
        }
        Ok(true)
    }
}
```

So, there are two main differences to this `PrefabData` compared the `Transform` example.

> The first difference is that the `add_to_entity` function now return a `Handle<A>`.

Reviewer: What happen to the returned Handle<A>? Since it's the PrefabLoaderSystem  that call the add_to_entity function, I may need to know what it does with it.

The second difference is that `load_sub_assets` is implemented, this is because we load
a sub asset. The `load_sub_assets` function here will do the actual loading, and morph the
internal representation to the `AssetPrefab::Handle` variant, so when `add_to_entity` runs later

> it will straight up use the internally stored `Handle`.

Reviewer: I understand that the load from the File variant in `add_to_entity` should never occurs?

### Special `PrefabData` implementations

There are a few special blanket implementations provided by the asset system:

* `Option<T>` for all `T: PrefabData`.
* Tuples of types that implemented `PrefabData`, up to a size of 20.

### Deriving `PrefabData` implementations

Amethyst supplies a derive macro for creating the `PrefabData` implementation for the following scenarios:

* Single `Component`
* Aggregate `PrefabData` structs which contain other `PrefabData` constructs, and optionally simple data `Component`s

In addition, deriving a `Prefab` requires that `amethyst::ecs::Entity` and

>  `amethyst:assets::{PrefabData, PrefabError, ProgressCounter}` are imported

Reviewer: ```suggestion
 `amethyst_assets::{PrefabData, PrefabError, ProgressCounter}` are imported
```

 and visible in the current scope. This is due to how Rust macros work.

An example of a single `Component` derive:

```rust,no_run,noplaypen
# #[macro_use] extern crate amethyst;
# #[macro_use] extern crate serde_derive;
# use amethyst::assets::{Asset, AssetStorage, Loader, Format, Handle, ProgressCounter, PrefabData, PrefabError};

> # use amethyst::ecs::{WriteStorage, ReadExpect, Read, Entity, error::Error as SpecsError, DenseVecStorage, Component};

Reviewer: I suggest to not hide imports of `Prefab` and `amethyst_assets::{PrefabData, PrefabError, ProgressCounter}` since this should not be forgotten.
